### PR TITLE
v0.15.0 changelog

### PR DESCRIPTION
**Pull request by chentsulin**
# 0.15.0 / 2018-07-18

`v0.15` is the second major version after we open sourced Bottender. In this version, we introduce a lot of helpful, community requested features based on [Messaging APIs v0.7](https://github.com/Yoctol/messaging-apis/releases/tag/v0.7.0).

- [new] add `context.requestContext`:

Express, Micro, Restify:

```js
context.requestContext; // { req, res }
```

Koa:

```js
context.requestContext; // ctx in koa
```

- [new] add more debug logs and key change (#239, #295), so you can run bots with following `DEBUG` env:

```sh
DEBUG=bottender:*
DEBUG=bottender:request
DEBUG=bottender:session:read
DEBUG=bottender:session:write
```

- [new] skip and show warning when calling send API in Messenger echo delivery read event (#306)
- [fix] deepClone when read from `MemoryCacheStore` (#235)
- [deps] Upgrade to Babel 7

### messenger

- [breaking] remove deprecated `MessengerContext` method - `sendQuickReplies`
- [new] support Messenger platform v2.4.
- [new] enable verifying graph API calls with `appsecret_proof` by default.
- [new] `context.getThreadOwner`

```js
const threadOwner = await context.getThreadOwner();
// {
//   app_id: '12345678910'
// }
```

- [new] add `pageId`, `gamePlay`, `brandedCamera`, `isRequestThreadControlFromPageInbox` getters to `MessengerEvent`

```js
context.event.pageId; // "<PAGE_ID>"

context.event.isRequestThreadControlFromPageInbox; // true

context.event.isGamePlay; //
context.event.gamePlay; //

context.event.isBrandedCamera; //
context.event.brandedCamera; //
```

- [new] implement Batch Mode to send multiple requests in one batch (up to 50 messages):

```js
const { isError613 } = require('messenger-batch');

new MessengerBot({
  // ...
  batchConfig: {
    delay: 1000,
    shouldRetry: isError613, // (#613) Calls to this api have exceeded the rate limit.
    retryTimes: 2,
  },
});
```

It will enable [message batching](https://github.com/Yoctol/messaging-apis/tree/master/packages/messaging-api-messenger#message-batching) functionality via [messaging-api-messenger](https://github.com/Yoctol/messaging-apis/tree/master/packages/messaging-api-messenger) under the hood.

- [deprecated] `sendAirlineFlightUpdateTemplate` has been renamed to `sendAirlineUpdateTemplate`.

### line

- [new] support LINE Flex Message with `replyFlex`, `pushFlex`, `sendFlex`:

```js
context.sendFlex('this is a flex', {
  type: 'bubble',
  header: {
    type: 'box',
    layout: 'vertical',
    contents: [
      {
        type: 'text',
        text: 'Header text',
      },
    ],
  },
  hero: {
    type: 'image',
    url: 'https://example.com/flex/images/image.jpg',
  },
  body: {
    type: 'box',
    layout: 'vertical',
    contents: [
      {
        type: 'text',
        text: 'Body text',
      },
    ],
  },
  footer: {
    type: 'box',
    layout: 'vertical',
    contents: [
      {
        type: 'text',
        text: 'Footer text',
      },
    ],
  },
  styles: {
    comment: 'See the example of a bubble style object',
  },
});
```

- [new] add `issueLinkToken` to `LineContext` (#245):

```js
const result = await context.issueLinkToken();
// {
//   linkToken: 'NMZTNuVrPTqlr2IF8Bnymkb7rXfYv5EY',
// }
```

- [new] add LINE `linkAccount` events support (#243):

```js
context.event.isAccountLink; // true
context.event.linkAccount;
// {
//   result: 'ok',
//   nonce: 'xxxxxxxxxxxxxxx',
// }
```

- [new] add `shouldBatch` option:

```js
new LineBot({
  // ...
  shouldBatch: true, // Default: false
});
```

When batching is enabled,

```js
context.replyText('Hi');
context.replyText('Hi');
context.replyText('Hi');
context.replyText('Hi');
context.replyText('Hi');
```

Those 5 messages will be sent in one API call, just like the below one.

```js
const { Line } = require('messaging-api-line');

context.reply([
  Line.createText('Hi'),
  Line.createText('Hi'),
  Line.createText('Hi'),
  Line.createText('Hi'),
  Line.createText('Hi'),
]);
```

- [new] add `sendMethod` option:

```js
new LineBot({
  // ...
  sendMethod: 'reply', // Default: 'push'
});
```

### telegram

- [breaking] Now context methods throw error when `ok` is `false` in Telegram:

```js
{
  ok: false,
  result: { /* ... */ }
}
```

### custom connector

- [new] pass merged query and body to handler:

```js
{
  ...query,
  ...body,
}
```

It's useful in custom connectors.

- [new] export Context from entry (#250)

```js
const { Context } = require('bottender');

class MyContext extends Context {
  //...
}
```

### CODE REVIEW

**Comment by tw0517tw** [comment]
LGTM

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,43 +1,254 @@
+# 0.15.0 / 2018-XX-XX
+
+`v0.15` is the second major version after we open sourced Bottender. In this version, we introduces a lot of helpful, community requested features based on [Messaging APIs v0.7](https://github.com/Yoctol/messaging-apis/releases/tag/v0.7.0).
```
Comment (kpman) [inline CHANGELOG.md]: we introduce a lot of helpful and community requested xxx

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,43 +1,254 @@
...
+```js
+new LineBot({
+  // ...
+  sendMethod: 'reply', // Default: 'push'
+});
+```
+
+### telegram
+
+- [breaking] Now context methods throws error when `ok` is `false` in Telegram:
```
Comment (wtflink) [inline CHANGELOG.md]: methods throw

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,43 +1,254 @@
...
+context.event.isRequestThreadControlFromPageInbox; // true
+
+context.event.isGamePlay; //
+context.event.gamePlay; //
+
+context.event.isBrandedCamera; //
+context.event.brandedCamera; //
+```
+
+- [new] implement Batch Mode to sends multiple requests in one batch (up to 50 messages):
```
Comment (wtflink) [inline CHANGELOG.md]: to send

**Pull request merged**